### ISSUE TITLE
[troika-three-text] IOS Safari Loading Issues

### ISSUE THREAD

**Issue by atlmtw** [protectwise/troika#88]
Hello!

First off I would like to say that this package has been WAY easier to use than other things in the past. So thank you very much for putting this out!

Anyways I've recently put this on a site I'm still working on and noticed some inconsistencies on mobile. Sometimes everything would load but other times things would be missing or incomplete. Screenshots below.

Would you happen to know what may be going on? Or if I am doing anything wrong?

The big font with "welcome" as the text is an .otf file (restgold.otf)
The small text with "Hi my name is..." as the text is a .ttf file (Raleway-Medium.ttf)
If you need the font files let me know!

Device: IPhone7
IOS: 14.1
Browser: Safari

Package Details:
    "three": "^0.122.0",
    "troika-three-text": "^0.35.0",
    "troika-three-utils": "^0.35.0",

![IMG_1509](https://user-images.githubusercontent.com/15266711/99882671-42ee7000-2bf0-11eb-8dc6-8b83ee8b52ba.PNG)
![IMG_1510](https://user-images.githubusercontent.com/15266711/99882672-43870680-2bf0-11eb-8275-6a4cc3d81ef2.PNG)
![IMG_1511](https://user-images.githubusercontent.com/15266711/99882673-43870680-2bf0-11eb-8247-c348cac9f5ee.PNG)
![IMG_1512](https://user-images.githubusercontent.com/15266711/99882674-441f9d00-2bf0-11eb-86df-25c892d9d4f5.PNG)

**Comment by lojjic** [comment]
Thanks for reporting this! You're not the first to describe issues like this in iOS Safari, but I wasn't able to reproduce it previously. I'll try with your site and hopefully I'll be able to reproduce it.

Are you using it via `drei` by chance? Or directly?

**Comment by atlmtw** [comment]
Hey @lojjic I'm using this package directly with three.js. And the reason I have three-utils is that I also use your createDerivedMaterial function in some cases.

Thanks for checking on this!

**Comment by lojjic** [comment]
I was able to reproduce the issue(s) loading your website on an iPhone 8. I haven't yet had time to start debugging but I just wanted to acknowledge that the bug is reproducible.

**Comment by RenaudROHLINGER** [comment]
@atlmtw @lojjic I had the same issue on safari on a project that was using multiple fonts. The only fix I found was to use a unique font on the whole website for this browser.

**Comment by strangerintheq** [comment]
+1 
multiple fonts per scene = some fonts never apper on iphone X, iphone SE

**Comment by lojjic** [comment]
I'm trying to dig into this issue, and I'm having trouble reproducing it now. Same iPhone 8 as before. I can't reproduce it on designsdalena.com anymore, but it's possible that site has changed to no longer use troika-three-text or worked around it in another way.

So I'm trying to create a minimal testcase using multiple fonts, and can't get it to fail:

https://uqgxq.csb.app/

Is anyone else able to reproduce this somewhere that I could use to test/debug?

**Comment by atlmtw** [comment]
Hey @lojjic 

To confirm, I did switch to something else. Although I do enjoy the experience of using troika more.

**Comment by lojjic** [comment]
@atlmtw Thanks for the confirmation. I don't suppose you have an old version of your site in source control that you'd be able to send me? I'm struggling to reproduce this and yours seemed like a reliable repro.

**Comment by amitrajput1992** [comment]
Hey @lojjic 
Any luck with this? I have been seeing this too when >1 fonts are used, on an iPhone (works on iPad and macOS). Weirdly this isn't replicable all the time, happens only 1-2/10 times. 

Btw, love the work :100:

**Comment by lojjic** [comment]
@amitrajput1992 No I still haven't been able to construct a reproducible test case. Do you have one I could use?

**Comment by lojjic** [comment]
@amitrajput1992 Thanks, I can indeed replicate the issue on that page. I'll see if I can debug from there and/or replicate it in a minimized local testcase.

> We know safari can be a pain with web workers

I've heard others say this too, and the usage of a worker is definitely a plausible culprit, but I haven't been able to find any details about known bugs with workers in Safari. Do you have specifics there you can share?

**Comment by amitrajput1992** [comment]
That sounds good. Let me know if I can help in any way. Meanwhile, I'll keep debugging and try to narrow down the issue on my end too. 


> I've heard others say this too, and the usage of a worker is definitely a plausible culprit, but I haven't been able to find any details about known bugs with workers in Safari. Do you have specifics there you can share?

I don't know about the exact issues here, partly because I can't find these anywhere (or people don't log these), but this is something that I've noticed while tinkering around. react360 formerly react-vr (abandoned now) used to runs the whole react code inside a web worker and the updates to the main thread were too damn slow. It would easily take a click action anywhere around 300ms to 500ms to propagate to my click listener in the worker thread.  
I maintain a repo that is a gif renderer for three, tried initially with offscreen canvas but the results were horrible and resulted in a blending of textures on consecutive frames. Moving the whole thing to the main thread improved it drastically.

**Comment by lojjic** [comment]
I think I should first focus on these artifacts, which don't always appear but sometimes do:

![Image from iOS](https://user-images.githubusercontent.com/138124/121586796-511ba680-c9f1-11eb-99e6-bbdc8b24cbab.jpg)

In those cases the layout and SDF generation is obviously completing and returning to the main thread, but some of the SDFs appear to have inside/outside fills reversed in places. I can see how that could happen if the path data for those characters is incomplete, like only having half its path segments.

If something is happening during font load where the font's arraybuffer is truncated or corrupted, I think that could possibly result in this observed symptom. Likewise it could also result in totally empty results if truncated at certain places.

I'll investigate that as a possibility (maybe XHR giving a partial arraybuffer response?) but first step is getting this into a reproducible testcase that I can modify and run locally.

**Comment by amitrajput1992** [comment]
That makes sense. arraybuffer corruption could be a culprit here. 
I saw another discussion about making this whole process run on the main thread. Is that on the roadmap?

**Comment by lojjic** [comment]
Running on the main thread is possible, but would result in a pretty crappy experience blocking the main thread for potentially several seconds. That should be a last resort only.

Has your test page changed? It seems like it's only using a single font for all the different text objects now, at least on iOS...? I still see garbled SDFs with that single font sometimes though, which implies this may be exacerbated by multiple fonts but isn't limited to that.

**Comment by amitrajput1992** [comment]
I did add a fallback for iOS devices to use only a single font for the entire app. This is my production env, so can't have broken things there. 
I'll create another example in my staging env and send out a link here asap.

It is troublesome to find out that this is still happening with a single font too :facepalm:

**Comment by lojjic** [comment]
Hmm it actually seems like I can only get the bug to occur with your new single font when connected to Safari devtools, and it's pretty consistently bugged then. Not sure if that gives us any clue or not.

**Comment by lojjic** [comment]
Well, a little more progress... I've verified that I'm able to force the same partial glyph artifacts seen above by truncating the path commands for individual glyphs. I haven't yet been able to determine whether that is due to the original font data being incomplete (I think that would cause only one partial glyph, not many, so it seems unlikely), or if something is causing the for-loops to exit early (this seems impossible, but maybe there's some strange JIT bug). 

Either way, I'm getting stuck with Safari's USB connection devtools not being able to set breakpoints in the relevant code. @amitrajput1992 Would it be at all possible to get your test app as source files I can build/run locally, so I can try swapping out the troika code to debug?

**Comment by amitrajput1992** [comment]
@lojjic While I won't be able to share the original app code, let me set up a storybook repo with a similar structure that I use in my production app and add a test render for this. Will send over the link shortly.

**Comment by amitrajput1992** [comment]
@lojjic 
I Have set up a basic repro with a similar structure as my production app here: https://github.com/amitrajput1992/r3f-experiments
I can replicate the same issue with this on my iPhone 11 and on BrowserStack. 
Also published the storybook here for easy access https://amitrajput1992.github.io/r3f-experiments

**Comment by amitrajput1992** [comment]
Hey @lojjic, I don't have a mac system with me right now, but I tested this on browserstack with local forwarding. Looks like the sdf data logged inside the web-worker vs main thread is different. It could be due to the serialization-deserialization process between thread communication, but not entirely sure. I'll keep debugging this further. 
You can try out crossbrowsertesting if safari dev-tools isn't working for you, they offer a 100min trial which might be helpful.

**Comment by munrocket** [comment]
> not being able to set breakpoints in the relevant code

Dumb suggestion to place console.log everywhere, since you guys are able to reproduce bug.

Picture on my iPhone 11:
![](https://habrastorage.org/webt/j1/ah/gf/j1ahgfmnq5xis6j_9txgfhi4d1g.jpeg)

**Comment by lojjic** [comment]
> Dumb suggestion to throw message from webworker after each line of code and console.log it

Not a dumb suggestion at all! And that's exactly what I was trying to do, but Safari's devtools crashes immediately for me when pointing to a local editable instance so I'm not even able to see the console.log output. :(

**Comment by amitrajput1992** [comment]
Are you trying to open a **localhost** url in the connected iPhone? How does port forwarding work in this case?

**Comment by lojjic** [comment]
> Are you trying to open a localhost url in the connected iPhone? How does port forwarding work in this case?

I'm accessing the dev server from the iPhone via local network IP address. I've also tried piping it through https://localhost.run. In both cases Safari devtools crashes as soon as I open it. The page itself renders just fine (though with the bug sometimes.)

**Comment by amitrajput1992** [comment]
I read on a few blogs that this can happen in 2 scenarios:
1. when the phone battery is at 100%
2. when debugging via network and not cable connection

This is a long running thread on similar lines, but no resolution https://developer.apple.com/forums/thread/92290

**Comment by munrocket** [comment]
>  but Safari's devtools crashes immediately for me when pointing to a local editable instance so I'm not even able to see the console.log output. :(

It is possible to substitute default console.log function with something like this
```js
console.log = (msg) => document.getElementById("my_ios_console").innerHTML += msg;
```
but you need to create that div on html page or from JS script
```html
<div id="my_ios_console" style="position: absolute; top:0; left: 0; background: white"></div>
```
this should show all console messages on main thread

**Comment by lojjic** [comment]
Thanks @munrocket, that could work, I'll give it a try.

**Comment by atlmtw** [comment]
Hey All,

Sorry I've been so away from this thread. Idk if this would help with debugging, but the recent Xcode 13 versions (in beta) simulator has been able to run my 3D localhost stuff fine! I ran into the issues y'all talked about before where it kept crashing with earlier versions.

**Comment by amitrajput1992** [comment]
@lojjic Any luck with this?

**Comment by lojjic** [comment]
> Any luck with this?

No, unfortunately, I've not been able to devote much time to this recently.

**Comment by strangerintheq** [comment]
any possibility to switch off web workers? just for check

**Comment by lojjic** [comment]
OMG @malulleybovo I returned from vacation and saw your findings and wow that was a wonderful surprise! 😃 Thank you so much for digging into this.

Just knowing that readShort is producing NaNs is a _huge_ step forward in maybe finally understanding this issue, which as you know I'd been totally stuck on for quite some time. It didn't help that I changed jobs and lost access to the iOS device I was using.

My next question would be can we determine _why_ the Typr._bin.t.int16[0] read is producing a NaN? It seems it must be getting an incorrect value in one of the buffers (either the font's `buff` or the utility `Typr._bin.t.buff`), but it would help to know exactly what the buff/uint8/int16 values are at that point in time versus what they should be.

The fact you can insert the console.log() there to avoid the bug is curious. I'm not sure if that indicates a race condition, or perhaps accessing the console takes it out of JIT mode. I'd hope for the former, as that sounds easier to detect and work around.

**Comment by lojjic** [comment]
@amitrajput1992 Hi, have you had a chance to test the alpha yet? I want to get this released and would love the extra validation. :)

**Comment by amitrajput1992** [comment]
@lojjic Hey, I just got the time to test this out. Looks like it's working now!!

Published the changes here: https://amitrajput1992.github.io/r3f-experiments/?path=/story/testers--text-tester

**Comment by lojjic** [comment]
I've released version `0.44.0` with the fix for this nasty bug. I'm so happy to finally have this fixed! Thank you everyone for the assistance, and especially @malulleybovo for digging deep where I was unable and finding the root cause. I'm very thankful! 🥳
